// @babylonjs/core/Particles/Node/Blocks/Update/basicSpriteUpdateBlock.js
import { RegisterClass } from "../../../../Misc/typeStore.js";
import { NodeParticleBlockConnectionPointTypes } from "../../Enums/nodeParticleBlockConnectionPointTypes.js";
import { NodeParticleBlock } from "../../nodeParticleBlock.js";
import { _ConnectAtTheEnd } from "../../../Queue/executionQueue.js";
/**
 * Block used to provide the basic update functionality for particle sprite index.
 */
export class BasicSpriteUpdateBlock extends NodeParticleBlock {
    /**
     * Create a new BasicSpriteUpdateBlock
     * @param name defines the block name
     */
    constructor(name) {
        super(name);
        this.registerInput("particle", NodeParticleBlockConnectionPointTypes.Particle);
        this.registerOutput("output", NodeParticleBlockConnectionPointTypes.Particle);
    }
    /**
     * Gets the particle component
     */
    get particle() {
        return this._inputs[0];
    }
    /**
     * Gets the output component
     */
    get output() {
        return this._outputs[0];
    }
    /**
     * Gets the current class name
     * @returns the class name
     */
    getClassName() {
        return "BasicSpriteUpdateBlock";
    }
    /**
     * Builds the block
     * @param state defines the current build state
     */
    _build(state) {
        const system = this.particle.getConnectedValue(state);
        const processSprite = (particle) => {
            state.particleContext = particle;
            state.systemContext = system;
            particle.updateCellIndex();
        };
        const spriteProcessing = {
            process: processSprite,
            previousItem: null,
            nextItem: null,
        };
        if (system._updateQueueStart) {
            _ConnectAtTheEnd(spriteProcessing, system._updateQueueStart);
        }
        else {
            system._updateQueueStart = spriteProcessing;
        }
        this.output._storedValue = system;
    }
}
RegisterClass("BABYLON.BasicSpriteUpdateBlock", BasicSpriteUpdateBlock);
